ci: Add a job and helper scripts to periodically scan for CVEs

Smjert opened the pull request:
- Add the third_party_libraries_cves_scanner.py and github_api.py helper scripts to scan for third party libraries CVEs, using the NIST NVD APIs and database, and a local third party libraries manifest which provides the metadata necessary for the scan.

  When new CVEs are found, and if the script is run with the --create_issues flag, new issues will be opened in the repository, if not already opened, to keep track of them.

- Add a validate_manifest_libraries_versions.py and manifest.py helper scripts to verify that the manifest of libraries is always with the correct format and that is up to date.

- Add a CI job running periodically once a day will use these scripts to validate the manifest, scan for CVEs and open issues.

- Validate the manifest during PRs too.

Closes #7690

directionless commented:
Reading the ToS for the API, it says:
 > Services which utilize or access the NVD API are asked to display the following notice prominently within the application: "This product uses the NVD API but is not endorsed or certified by the NVD."

While we're not a service, I think it would be reasonable for us to talk about how the CVE scanning works in the readme (or docs?) and toss that in

directionless commented on .github/workflows/cve_scan_runner.yml:
```diff
@@ -0,0 +1,36 @@
+      - name: Scan CVEs and open issues
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        # NIST_API_KEY: ${{ secrets.NIST_API_KEY }}
```
This is https://nvd.nist.gov/developers correct?

I have registered osquery for an API key. Added it to 1password, and added it to github secrets as `NVD_API_KEY`

```suggestion
        NIST_API_KEY: ${{ secrets.NVD_API_KEY}}
```

directionless commented on tools/ci/scripts/cve/osquery/github_api.py:
```diff
@@ -0,0 +1,186 @@
+    def makeGetRequest(self, url, params: dict):
+        time_passed = (time.time_ns() - self.last_request_time) / (1000 * 1000)
+
+        # Limit at 1 request every 5 seconds
```
Where's this limit from? It seems slow?

directionless replied:
Answering myself, it's:
> When using GITHUB_TOKEN, the rate limit is 1,000 requests per hour per repository. For requests to resources that belong to an enterprise account on GitHub.com, GitHub Enterprise Cloud's rate limit applies, and the limit is 15,000 requests per hour per repository.

So evert 5s feels reasonable

directionless commented on tools/ci/scripts/cve/osquery/github_api.py:
```diff
@@ -0,0 +1,186 @@
+        # or something that ignores the CVEs have been created.
+        since = datetime.utcnow() - timedelta(days=180)
+
+        params = {
```
I can't tell whether or not this is correctly searching, or leaving it for client side. If I read the python right, it looks like it's being submitted?

Either way, does `https://api.github.com/search/issues?q=is:issue%20repo:osquery/osquery%20label:security` get closer?

Smjert replied:
The `since` is passed to the API call just below, in the `params` dictionary.

About the time limit, I partially misremembered during the office hours but, the first reason is indeed a limit on the issues being retrieved, but only on the CVE issues themselves. Either way, as you can see I can only retrieve 100 issues every 5 seconds. Hopefully it won't grow that quickly?

The second convenience that having a separate JSON file that provides the ignored CVEs, is that it partially stores a state, a point in time, that can be checked at any time.
It might not be as useful, but suppose the current version of osquery is 5.9.0 and I'm downloading the source code for 5.7.0 and wanted to check how was the state of the CVEs at that time, I could've re-run the script and had the view. You can also check the state down to the specific commit.

The issues instead always show the last state. That been said, I'm not saying it's a vital feature, just thought it was useful. We can emulate that in other ways I guess, by creating a list of the CVEs found, ignored and resolved that we report somewhere.

directionless replied:
The idea of having a snapshot in the release is a good one.

If we felt strongly about using Issues as a tracking DB, we could capture the state into a release file. 

I'm not totally sure where I fall.

What you have is mostly simple to implement. But I'm a little worried that needing to update the manifest will be cumbersome. But, the counter point is that updating dependancies is also cumbersome. So maybe that's fine. 

Where are you envisioning discussion about a CVE goes? Into the github issue?

Smjert replied:
Just to compare and contrast a little more:

So, the Issues way is a bit more automatic, meaning, one closes a CVE and it's automatically regarded as ignored, without review. 
On the contrary the manifest way is less automatic, but it means that a review happened before closing an issue; if an issue is closed and no review happened, it will be reopened, much like not resolving a CVE we should not ignore, because no library update has been done.

Obviously as mentioned in the office hours, someone may close the issue without doing the PR, but at that point, opening a new issue with the same CVE is a signal that the procedure has not been followed. I could maybe even list closed CVEs and find one to reopen, if there is, instead of opening a duplicate.
I feel like deciding to ignore a CVE is a pretty important decision, so having a PR process with it was additionally my way of ensuring that there were enough eyes on it.

Finally, another issue comes to my mind (which I thought about when developing but did not immediately recorded). Hopefully not but the script may have bugs, or we want to change the format of the title; ignoring closed and old CVE issues was a way to permit change to happen there.
I felt like changing a JSON file was much quicker than going back to change all the old issues, or start adding support for different ways to parse the issue titles.

> Where are you envisioning discussion about a CVE goes? Into the github issue?
Ideally yes; as a "forum" and to keep track of the discussion I think that's best. We can obviously further discuss things on Slack, but the issues should at least capture a clear reasoning of why a CVE issue has been ignored/closed.

directionless commented on tools/ci/scripts/cve/third_party_libraries_cves_scanner.py:
```diff
@@ -0,0 +1,383 @@
+                github_api.createIssue(
+                    f"Library {library_name} has vulnerability {cve.name}",
+                    cve.description,
+                    ["cve-%s" % cve.severity.lower(), "libraries"],
```
Are these the labels? If so, I'd recommend:
1. We add `security`
2. maybe split `cve` and `severity` in two

These should make it more searchable

directionless commented on tools/ci/scripts/cve/third_party_libraries_cves_scanner.py:
```diff
@@ -0,0 +1,383 @@
+
+            try:
+                github_api.createIssue(
+                    f"Library {library_name} has vulnerability {cve.name}",
```
This is the subject? We should be the CVE number in there. (if it's not already in cve.name)

Smjert replied:
Yes the `cve.name` is already the CVE full id. Please check the issues created in this example repository: https://github.com/Smjert/osquery-cve-tmp/issues

directionless commented on tools/ci/scripts/cve/third_party_libraries_cves_scanner.py:
```diff
@@ -0,0 +1,383 @@
+            try:
+                github_api.createIssue(
+                    f"Library {library_name} has vulnerability {cve.name}",
+                    cve.description,
```
Do we have a URL? Can we include it somewhere?

directionless commented on .github/workflows/cve_scan_runner.yml:
```diff
@@ -0,0 +1,36 @@
+  workflow_dispatch:
+
+  schedule:
+    - cron: "0 0 * * *"
```
```suggestion
    - cron: "17 23 * * *"
```

Practically speaking, this tends to be a busy time for cronjobs. Consider something not on the 0s.

directionless commented on .github/workflows/cve_scan_runner.yml:
```diff
@@ -0,0 +1,36 @@
+
+      - name: Install python pre-requisites
+        run: |
+          pip3 install nvdlib==0.6.0 pygit2
```
would this be cleaner in a requirements.txt?

directionless reviewed (commented):
Some comments, but we'd also talked about whether we'll be able to use github issues directly as a DB for this.

If it would help, we could create a dedicated repo to hold the issues. Though I'm not sure it would be a win over figuring out something uniqueish to search on

Smjert commented:
@directionless I have updated as you suggested. For the new labels and URLs please have a look at https://github.com/Smjert/osquery-cve-tmp/issues/35 as an example.

I've also added a wiki page named cve-scan.md under the development category with some documentation.

Smjert commented:
I've frozen the pygit2 version, since newer version were expecting newer versions of pip and didn't necessarily want to always do an upgrade vs use what we have on the system.

I've also removed a stray debug print.

Smjert commented:
Rebased against master to pick up the CI python fix.

directionless reviewed (approved):
While I'm not sure this is perfect, I think we should merge and iterate.

Sorry for dragging my feet on it, I think it will be a huge boon.
